BinaryVault/contracts/BinaryVault.sol

```diff
@@ -0,0 +1,1580 @@
+        // Transfer underlying token from user to the vault
+        underlyingToken.safeTransferFrom(user, address(this), amount);
+
+        uint256 newShares = 0;
+        if (totalShareSupply > 0) {
+            newShares = (amount * totalShareSupply) / underlyingTokenBalance;
+        } else {
+            newShares = amount;
+            require(newShares > MINIMUM_LIQUIDITY, "Insufficient amount");
+        }
+
+        shareBalances[tokenId] = newShares;
+        
+        totalShareSupply += newShares;
+
+        emit NewLiquidityAdded(user, tokenId, amount, newShares);
+    }
+
+    /**
+    * @dev Add liquidity. Burn existing token, mint new one.
+    * @param user Receipt for share
+    * @param amount Underlying token amount
+    * @param tokenId nft id to be added liquidity. This should be existing id.
+    */
+    function addLiquidityPosition(address user, uint256 amount, uint256 tokenId)
+        external
+        override
+        whenNotPaused
+        nonReentrant
+    {
+        if (user == address(0)) revert("ZERO_ADDRESS()");
+        if (amount == 0) revert("ZERO_AMOUNT()");
+        require(_exists(tokenId), "Non exists token");
+        require(ownerOf(tokenId) == user, "Not owner");
+        require(user == msg.sender || isApprovedForAll(user, msg.sender), "Not approved or owner");
+
+        uint256 underlyingTokenBalance = underlyingToken.balanceOf(address(this));
+        // Transfer underlying token from user to the vault
+        underlyingToken.safeTransferFrom(user, address(this), amount);
+
+        uint256 newShares = 0;
+        if (totalShareSupply > 0) {
+            newShares = (amount * totalShareSupply) / underlyingTokenBalance;
+        } else {
+            newShares = amount;
+            require(newShares > MINIMUM_LIQUIDITY, "Insufficient amount");
+        }
+
+        uint256 currentShare = shareBalances[tokenId];
+        // Burn existing one
+        _burn(tokenId);
+        delete shareBalances[tokenId];
+        // Mint new one
+        uint256 newTokenId = _nextTokenId();
+        _mint(user, 1);
+
+        shareBalances[newTokenId] = currentShare + newShares;
+        
+        totalShareSupply += newShares;
+
+        // TODO Update event
+        emit LiquidityAdded(user, tokenId, newTokenId, amount, newShares);
+    }
+
+    /**
+    * @dev Merge tokens into one, Burn existing ones and mint new one
+    * @param user receipent
+    * @param tokenIds Token ids which will be merged
+    */
+    function mergePositions(address user, uint256[] memory tokenIds) public {
+        require(user != address(0), "Invalid user");
+        require(user == msg.sender || isApprovedForAll(user, msg.sender), "Non owner or approved");
+
+        uint256 shareAmounts = 0;
+        for (uint256 i; i < tokenIds.length; i = i + 1) {
+            uint256 tokenId = tokenIds[i];
+            require(_exists(tokenId), "Non exists token");
```

Comment:
Please change the requirement message types
require(_exists(tokenId), "BinaryVault: Non exists token");